# 1. Introduction

In this project, we'll work with a data set of submissions to popular technology site Hacker News.

Hacker News is a site started by the startup incubator Y Combinator, where user-submitted stories (known as "posts") are voted and commented upon, similar to reddit. Hacker News is extremely popular in technology and startup circles, and posts that make it to the top of Hacker News' listings can get hundreds of thousands of visitors as a result.

The relevant dataset can be found at https://www.kaggle.com/hacker-news/hacker-news-posts (accessed on April 21, 2021).

Note that it has been reduced from almost 300,000 rows to approximately 20,000 rows by removing all submissions that did not receive any comments, and then randomly sampling from the remaining submissions. Below are descriptions of the columns:

    1. id: The unique identifier from Hacker News for the post
    
    2. title: The title of the post
    
    3. url: The URL that the posts links to, if the post has a URL
    
    4. num_points: The number of points the post acquired, calculated as the total number of upvotes minus the total number of downvotes
    
    5. num_comments: The number of comments that were made on the post
    
    6. author: The username of the person who submitted the post
    
    7. created_at: The date and time at which the post was submitted

We're specifically interested in posts whose titles begin with either Ask HN or Show HN. Users submit Ask HN posts to ask the Hacker News community a specific question. Below are a couple examples:

Ask HN: How to improve my personal website?
Ask HN: Am I the only one outraged by Twitter shutting down share counts?
Ask HN: Aby recent changes to CSS that broke mobile?
Likewise, users submit Show HN posts to show the Hacker News community a project, product, or just generally something interesting. Below are a couple of examples:

Show HN: Wio Link  ESP8266 Based Web of Things Hardware Development Platform'
Show HN: Something pointless I made
Show HN: Shanhu.io, a programming playground powered by e8vm
We'll compare these two types of posts to determine the following:

Do Ask HN or Show HN receive more comments on average?
Do posts created at a certain time receive more comments on average?
Let's start by importing the libraries we need and reading the data set into a list of lists.

In [1]:
from csv import reader
opened_file = open("hacker_news.csv")
read_file = reader(opened_file)
hn = list(read_file)

for i in range(0, 5):
    print(hn[i])

['id', 'title', 'url', 'num_points', 'num_comments', 'author', 'created_at']
['12579008', 'You have two days to comment if you want stem cells to be classified as your own', 'http://www.regulations.gov/document?D=FDA-2015-D-3719-0018', '1', '0', 'altstar', '9/26/2016 3:26']
['12579005', 'SQLAR  the SQLite Archiver', 'https://www.sqlite.org/sqlar/doc/trunk/README.md', '1', '0', 'blacksqr', '9/26/2016 3:24']
['12578997', 'What if we just printed a flatscreen television on the side of our boxes?', 'https://medium.com/vanmoof/our-secrets-out-f21c1f03fdc8#.ietxmez43', '1', '0', 'pavel_lishin', '9/26/2016 3:19']
['12578989', 'algorithmic music', 'http://cacm.acm.org/magazines/2011/7/109891-algorithmic-composition/fulltext', '1', '0', 'poindontcare', '9/26/2016 3:16']


# 2. Removing Headers from a List of Lists

As the first row od the dataset contains the column headers, we will remove that row as follows:

# 2. Removing Headers from a List of Lists

In [2]:
hn = hn[1:]

for i in range(0, 5):
    print(hn[i])

['12579008', 'You have two days to comment if you want stem cells to be classified as your own', 'http://www.regulations.gov/document?D=FDA-2015-D-3719-0018', '1', '0', 'altstar', '9/26/2016 3:26']
['12579005', 'SQLAR  the SQLite Archiver', 'https://www.sqlite.org/sqlar/doc/trunk/README.md', '1', '0', 'blacksqr', '9/26/2016 3:24']
['12578997', 'What if we just printed a flatscreen television on the side of our boxes?', 'https://medium.com/vanmoof/our-secrets-out-f21c1f03fdc8#.ietxmez43', '1', '0', 'pavel_lishin', '9/26/2016 3:19']
['12578989', 'algorithmic music', 'http://cacm.acm.org/magazines/2011/7/109891-algorithmic-composition/fulltext', '1', '0', 'poindontcare', '9/26/2016 3:16']
['12578979', 'How the Data Vault Enables the Next-Gen Data Warehouse and Data Lake', 'https://www.talend.com/blog/2016/05/12/talend-and-Â\x93the-data-vaultÂ\x94', '1', '0', 'markgainor1', '9/26/2016 3:14']


# 3. Extracting Ask HN and Show HN Posts 

Now that we've removed the headers from hn, we're ready to filter our data. Since we're only concerned with post titles beginning with Ask HN or Show HN, we'll create new lists of lists containing just the data for those titles.

To do that, we will:

    1. Create three empty lists called ask_posts, show_posts, and other_posts.
    2. Loop through each row in hn.
    3. Assign the title in each row to a variable named title; and
    4. Implement the following steps:
        a) If the lowercase version of title starts with ask hn, append the row to ask_posts.
        b) Else if the lowercase version of title starts with show hn, append the row to show_posts.
        c) Else append to other_posts.
    5. Check the number of posts in ask_posts, show_posts, and other_posts.

In [3]:
ask_posts = []
show_posts = []
other_posts = []

for row in hn:
    title = row[1]
    if title.lower().startswith("ask hn"):
        ask_posts.append(row)
    elif title.lower().startswith("show hn"):
        show_posts.append(row)
    else: other_posts.append(row)

print(f"number of ask posts: {len(ask_posts)}")
print(f"number of show posts: {len(show_posts)}")
print(f"number of other posts: {len(other_posts)}")

number of ask posts: 9139
number of show posts: 10158
number of other posts: 273822


Below are the first five rows in the ask_posts list of lists:

In [4]:
for i in range(0, 5):
    print(ask_posts[i])

['12578908', 'Ask HN: What TLD do you use for local development?', '', '4', '7', 'Sevrene', '9/26/2016 2:53']
['12578522', 'Ask HN: How do you pass on your work when you die?', '', '6', '3', 'PascLeRasc', '9/26/2016 1:17']
['12577908', 'Ask HN: How a DNS problem can be limited to a geographic region?', '', '1', '0', 'kuon', '9/25/2016 22:57']
['12577870', 'Ask HN: Why join a fund when you can be an angel?', '', '1', '3', 'anthony_james', '9/25/2016 22:48']
['12577647', 'Ask HN: Someone uses stock trading as passive income?', '', '5', '2', '00taffe', '9/25/2016 21:50']


Below are the first five rows in the show_posts list of lists:

In [5]:
for i in range(0, 5):
    print(show_posts[i])

['12578335', 'Show HN: Finding puns computationally', 'http://puns.samueltaylor.org/', '2', '0', 'saamm', '9/26/2016 0:36']
['12578182', 'Show HN: A simple library for complicated animations', 'https://christinecha.github.io/choreographer-js/', '1', '0', 'christinecha', '9/26/2016 0:01']
['12578098', 'Show HN: WebGL visualization of DNA sequences', 'http://grondilu.github.io/dna.html', '1', '0', 'grondilu', '9/25/2016 23:44']
['12577991', 'Show HN: Pomodoro-centric, heirarchical project management with ES6 modules', 'https://github.com/jakebian/zeal', '2', '0', 'dbranes', '9/25/2016 23:17']
['12577142', 'Show HN: Jumble  Essays on the go #PaulInYourPocket', 'https://itunes.apple.com/us/app/jumble-find-startup-essay/id1150939197?ls=1&mt=8', '1', '1', 'ryderj', '9/25/2016 20:06']


# 4. Calculating the Average Number of Comments for Ask HN and Show HN Posts

Below we will determine if ask posts or show posts receive more comments on average.

To that end, we will:

    1. Find the total number of comments in ask posts and assign it to total_ask_comments.

    2. Use a for loop to iterate over the ask posts.

    3. Compute the average number of comments on ask posts and assign it to avg_ask_comments.

    4. Find the total number of comments in show posts and assign it to total_show_comments.

    5. Use a for loop to iterate over the show posts.

    6. Compute the average number of comments on show posts and assign it to avg_show_comments.
    Print avg_show_comments.

In [6]:
total_ask_comments = 0

for row in ask_posts:
    comments = int(row[4])
    total_ask_comments += comments

avg_ask_comments = total_ask_comments / len(ask_posts)

print(f"The average of comments per ask post is {avg_ask_comments}.")

total_show_comments = 0

for row in show_posts:
    comments = int(row[4])
    total_show_comments += comments

avg_show_comments = total_show_comments / len(show_posts)

print(f"The average of comments per show post is {avg_show_comments}.")

The average of comments per ask post is 10.393478498741656.
The average of comments per show post is 4.886099625910612.


As shown above, on average, ask posts receive more comments (app. 10.39 per post) than show comments (app. 4.89 per post).

# 5. Finding the Amount of Ask Posts and Comments by Hour Created

Above we determined that, on average, ask posts receive more comments than show posts. Since ask posts are more likely to receive comments, we'll focus our remaining analysis just on these posts.

Next, we'll determine if ask posts created at a certain time are more likely to attract comments. We'll use the following steps to perform this analysis:

    1. Calculate the amount of ask posts created in each hour of the day, along with the number of comments received.
    2. Calculate the average number of comments ask posts receive by hour created.

To tackle the first step — calculating the amount of ask posts and comments by hour created. We'll use the datetime module to work with the data in the created_at column.

To that end, we will:

    1. Import the datetime module as dt.
    
    2. Create an empty list and assign it to result_list. This will be a list of lists.
    
    3. Iterate over ask_posts and append to result_list a list with two elements:
        (a) The first element shall be the column created_at.
        (b)The second element shall be the number of comments of the post.

    4. Create two empty dictionaries called counts_by_hour and comments_by_hour.
    
    5. Loop through each row of result_list.
    
    6. Extract the hour from the date, which is the first element of the row.
    
    7. Use the datetime.strptime() method to parse the date and create a datetime object.
    
    8. Use the string we want to parse as the first argument and a string that specifies the format as the second argument.


In [21]:
import datetime as dt

result_list = []

for row in ask_posts:
    created_at = row[6]
    num_comments = row[4]
    elements = [created_at, num_comments]
    result_list.append(elements)
    
counts_by_hour = {}

comments_by_hour = {}

for row in result_list:
    created_at = dt.datetime.strptime(row[0], "%m/%d/%Y %H:%M")
    created_at_hour = created_at.strftime("%H")
    num_comments = int(row[1])
    if created_at_hour not in counts_by_hour:
        counts_by_hour[created_at_hour] = 1
        comments_by_hour[created_at_hour] = num_comments
    else:
        counts_by_hour[created_at_hour] += 1
        comments_by_hour[created_at_hour] += num_comments

# 6. Calculating the Average Number of Comments for Ask HN Posts by Hour


Above, we created two dictionaries:

    counts_by_hour: contains the number of ask posts created during each hour of the day.
    
    comments_by_hour: contains the corresponding number of comments ask posts created at each hour received.
    
Next, we'll use these two dictionaries to calculate the average number of comments for posts created during each hour of the day.

To do that, we will reate a list of lists containing the hours during which posts were created and the average number of comments those posts received as follows:

    1. Calculate the average number of comments per post for posts created during each hour of the day; and
    
    2. Assign the result to a variable named avg_by_hour. Display the results.

In [37]:
comments = []

avg_by_hour = []
    
for key in counts_by_hour:
    avg_comments = (comments_by_hour[key] / counts_by_hour[key])
    avg_arr = [key, avg_comments]
    avg_by_hour.append(avg_arr)
    
print(avg_by_hour)

[['02', 11.137546468401487], ['01', 7.407801418439717], ['22', 8.804177545691905], ['21', 8.687258687258687], ['19', 7.163043478260869], ['17', 9.449744463373083], ['15', 28.676470588235293], ['14', 9.692007797270955], ['13', 16.31756756756757], ['11', 8.96474358974359], ['10', 10.684397163120567], ['09', 6.653153153153153], ['07', 7.013274336283186], ['03', 7.948339483394834], ['23', 6.696793002915452], ['20', 8.749019607843136], ['16', 7.713298791018998], ['08', 9.190661478599221], ['00', 7.5647840531561465], ['18', 7.94299674267101], ['12', 12.380116959064328], ['04', 9.7119341563786], ['06', 6.782051282051282], ['05', 8.794258373205741]]


# 7. Sorting and Printing Values from a List of Lists

Although we now have the results we need, this format makes it hard to identify the hours with the highest values. The following will sort a list of lists and print the five highest values in a format that's easier to read:

In [55]:
swap_avg_by_hour = []

for arr in avg_by_hour:
    swap = [arr[1], arr[0]]
    swap_avg_by_hour.append(swap)

sorted_swap = sorted(swap_avg_by_hour, reverse=True)

print("Top 5 Hours for Ask Posts Comments:")

template = "{}: {:.2f} average comments per post."

for arr in sorted_swap[:5]:
    percentage = arr[0]
    hour = arr[1]
    hour = dt.datetime.strptime(arr[1], "%H")
    hour = hour.strftime("%H:%M")
    print(template.format(hour, percentage))

Top 5 Hours for Ask Posts Comments:
15:00: 28.68 average comments per post.
13:00: 16.32 average comments per post.
12:00: 12.38 average comments per post.
02:00: 11.14 average comments per post.
10:00: 10.68 average comments per post.
